Port `cat` kernel to structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #73351
* #73350
* #69607
* #69606
* __->__ #68640

Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

ysiraichi committed:
Port `cat` kernel to structured kernels.

Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -27,6 +28,147 @@
+  bool all_contiguous = true;
+  bool all_same_dtype = true;
+  bool all_same_sizes_and_stride = true;
+  auto memory_format = cat_compute_output_memory_format(tensors);
```
CPU and CUDA kernels computed the memory format in different ways. Here, I adopted how CUDA used to do it. Not sure if this is the best way to go at it. Any thoughts?

bdhirsh replied:
cc @ngimel for another opinion. I found the PR that [added support for cuda](https://github.com/pytorch/pytorch/pull/34526), and a [bit later for cpu](https://github.com/pytorch/pytorch/pull/39448). There's no explicit mention of why they differ, so, going with the way cuda does it now seems reasonable to me.

bdhirsh replied:
@mruberry helped point me to some discussion about this: https://github.com/pytorch/pytorch/pull/62560#issuecomment-905923630.

It sounds like making cuda's behavior the general behavior is the right move in this PR. We should also add a test (if one doesn't exist already), that confirms that memory_format behavior is the same across cpu and cuda.

@mruberry also pointed out another thing we need to fix in this PR: see https://github.com/pytorch/pytorch/issues/64709. Apparently, there parts of the codebase that call `cat` and expect `cat(out=...)` to resize the output tensor. That'll start printing warnings now that `cat` is structured. We should fix all of the places in our codebase that call `cat(out=...)` and expect the resize to happen, that way users don't start seeing warnings that they can't fix.

bdhirsh replied:
Existing cat memory_format test is here: https://github.com/pytorch/pytorch/blob/1d269e8c15da8bd4a3bfcfbf44ae27ec36e9735b/test/test_tensor_creation_ops.py#L721

AnirudhDagar replied:
Thanks @ysiraichi, indeed the CUDA and CPU memory format calculation is inconsistent. I opened issue #63998 (talking exactly about this) earlier and wanted to work on a fix for that but never got around that because of some new responsibilities until earlier this week. I just came across this PR by chance and I see you have already incorporated the change.

So, with this PR, #63998 should also be fixed.

@bdhirsh when adding the existing cat memory format test, I kept the different behaviour of memory formats in mind. It will now need slight modification with your update making CPU and CUDA consistent, which should be quite straightforward.

ysiraichi commented on test/test_tensor_creation_ops.py:
```diff
@@ -626,16 +626,22 @@ def test_cat_out(self, device):
         with self.assertRaisesRegex(
-                RuntimeError, r"unsupported operation:.* input tensor 0"):
+                RuntimeError,
+                r"unsupported operation: some elements of the input tensor and "
```
Error messages changed here, because I'm using `at::assert_no_overlap` function. Should I revert that to show better error messages?

bdhirsh replied:
This seems reasonable. The previous error message was also related to overlap:
```
>>> torch.cat([x, y], out=x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: 0unsupported operation: the input tensors cannot refer to any of the output memory locations. Found overlap in input tensor 0
```

ysiraichi replied:
Yeah. I was a bit worried that not having the number of the input tensor that had an overlap would be bad. :)

ysiraichi commented on test/test_tensor_creation_ops.py:
```diff
@@ -751,7 +756,7 @@ def test_cat_out_memory_format(self, device):
 
         self.assertTrue(res2_cuda.is_contiguous(memory_format=torch.contiguous_format))
-        self.assertTrue(res2_cpu.is_contiguous(memory_format=torch.channels_last))
+        self.assertTrue(res2_cpu.is_contiguous(memory_format=torch.contiguous_format))
```
Since I'm using only one method for inferring the memory format, CPU and CUDA behave consistently.

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -7730,18 +7737,6 @@
-    CPU: _cat_out_cpu
-    CUDA: cat_out_cuda
-    QuantizedCPU: cat_out_quantized_cpu
-
```
@zou3519 do you happen to know the history behind why we had both an `at::cat` and `at::_cat`? I couldn't dig up much of a reason from git blame. Although it seems usefully to try to kill it as part of this structured kernel port.

zou3519 replied:
I don't know why we have both. I agree if we can kill `_cat` we should; there doesn't seem to be a need for both `_cat` and `cat` to exist

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -999,12 +999,19 @@
+  structured_delegate: cat.out
   dispatch:
-    CompositeExplicitAutograd: cat
+    SparseCPU, SparseCUDA: cat_sparse
```
cleaning up the sparse logic :)

bdhirsh replied:
I guess for consistency we'd like to have an out= variant for sparse kernels, but it looks like that's not true today. It's probably not necessary to try to fix that in this PR.

bdhirsh commented on test/test_tensor_creation_ops.py:
```diff
@@ -739,8 +745,7 @@ def test_cat_out_memory_format(self, device):
         # Case 2: if out= is not the correct shape then the output it is resized internally
-        # - For the CPU variant the memory format is that of the first tensor
-        # - For the CUDA variant it only propagates memory format if all the tensors have
+        # - For both CPU and CUDA variants, it only propagates memory format if all the tensors have
```
@ngimel, is this change in memory format propagation on CPU an issue? It seems useful to clean this logic up so it's backend-agnostic, but it sounds mildly BC-breaking.

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -432,15 +471,15 @@ static Tensor cat_sparse(TensorList tensors, int64_t dim) {
-        native::cat(idxs_pieces, 1),
-        native::cat(vals_pieces),
+        at::cat(idxs_pieces, 1),
+        at::cat(vals_pieces),
```
extra dispatcher trip seems ok to me, since we're re-using this kernel for SparseCPU + SparseCUDA

bdhirsh commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -94,130 +239,52 @@ std::vector<Tensor> broadcast_tensors(TensorList tensors) {
     auto result_stride_bytes = result.stride(dim) * elementSize(result.scalar_type());
 
     auto iter = TensorIteratorConfig()
-      .set_check_mem_overlap(false)  // Already checked above
```
"Already checked above" still applies, right? You're doing the overlap check explicitly in the meta function

ysiraichi replied:
I thought leaving the comment would be a little confusing, since we do it in META (not "above", in the same function). Maybe I could just add: "Already checked in the META function". What do you think?

bdhirsh replied:
That sounds good!

bdhirsh commented:
It would also be nice to benchmark and see that `cat()` got faster thanks to `ITensorList`. I'm trying to think of the best way to do that though... you would need to exercise a boxed fallback in eager mode that runs before hitting the `cat` kernel. A few ways to do that are:
- Try to use an existing boxed fallback? Conjugate might work, which has a boxed fallback registered
- Make a branch where you register a boxed fallback to a new key, and turn the key on unconditionally. You can do this by adding the key to the default include set [here](https://github.com/pytorch/pytorch/blob/d026057bb32db256289d85f0a63d32922c6a5fcc/c10/core/DispatchKeySet.h#L232).

Let me know what you think

ysiraichi commented:
I think benchmarking it would be nice, indeed. That said, I'm not sure how to do it. It will probably take some time...

bdhirsh commented:
> I think benchmarking it would be nice, indeed. That said, I'm not sure how to do it. It will probably take some time...

Let me know how I can help! If you write a no-op boxed fallback, it would probably look something like this:
```
void noop_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
  // just redispatch directly to the CPU dispatch key. That will perform the unboxing logic, and call your unboxed cat kernel.
  // That should force the conversion from `c10::List<Tensor>` -> `ArrayRef<Tensor>`  before your PR.
  op.redispatchBoxed(c10::DispatchKeySet(c10::DispatchKey::CPU), stack);
}

TORCH_LIBRARY_IMPL(_, NOOP_KEY, m) {
  m.fallback(torch::CppFunction::makeFromBoxedFunction<&noop_fallback>());
}
```

And then add a NOOP key anywhere with higher priority than the CPU key in DispatchKey.h, and add it to the default_include keyset [here](https://github.com/pytorch/pytorch/blob/3d358a76783acef87f282cc0fb20818eeb4c133c/c10/core/DispatchKeySet.h#L232), so that the benchmark you run will always hit your new kernel. We're also... out of dispatch keys currently 😛 you could deal with that by deleting one of the PrivateUse keys or something in the branch. There are some tutorials on using `torch.benchmark` utils here: https://pytorch.org/tutorials/recipes/recipes/benchmark.html.

AnirudhDagar reviewed (commented):
@ysiraichi I left a small comment about the memory format changes. Thanks for this PR!

ysiraichi commented:
@bdhirsh I've been thinking... do we expect performance to be improved because of `ITensorList`?
- I still haven't changed the dispatcher API. In other words only structured kernels are aware of `ITensorList` (planning to do it in a later PR). 
- `IValue::toTensorVector()` is still called by the dispatcher

bdhirsh commented:
> I've been thinking... do we expect performance to be improved because of ITensorList?

> I still haven't changed the dispatcher API. In other words only structured kernels are aware of ITensorList (planning to do it in a later PR).
> IValue::toTensorVector() is still called by the dispatcher

Ah ok, I forgot about that. Just to clarify, is the problem that the templated unboxing logic right now will continue to hit [this code path](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h#L354)? I guess we also need to tell the unboxing logic how to "unbox an ivalue into an ITensorList" (which should be cheap, but isn't happening yet).

ysiraichi commented:
> Just to clarify, is the problem that the templated unboxing logic right now will continue to hit this code path? 

Exactly!

> I guess we also need to tell the unboxing logic how to "unbox an ivalue into an ITensorList" (which should be cheap, but isn't happening yet).

Yes! But, I will have to change the dispatch API, first.

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi commented on torch/utils/data/_utils/collate.py:
```diff
@@ -63,6 +63,8 @@ def default_collate(batch):
             numel = sum(x.numel() for x in batch)
             storage = elem.storage()._new_shared(numel)
             out = elem.new(storage)
+            # Resize 'out' before passing it to 'stack'
+            out = out.resize_((len(batch), *elem.shape))
```
After this change, the CI is now passing. In other words, I think `cat` and its variants don't seem to be raising warnings internally anymore. Thus, I will remove the `expectedFailure` from all of them: `cat`, `stack`, `hstack`, `vstack`, and `dstack`.

krshrimali replied:
Hi, @ysiraichi - I think you can also remove the skip from `column_stack` in this PR after this change since it also seems to be calling `hstack`. If you already had it noted, please ignore. :) Thank you for the PR.

ngimel commented:
Since `cat` is a very common operation (most popular pytorch operation), we need `cat` benchmarks to make sure we didn't regress overheads. Can you do benchmarks with timings and instruction counts, in particular for overhead-dominated cases (catting many small tensors)

ysiraichi commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -27,6 +28,156 @@
+  }
+
+  if (raises_warning) {
+    TORCH_CHECK(false, "cat_out(): bad output sizes raises warning");
```
This is just something to catch if `cat` is raising warnings internally. I will remove it from the final version.

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Fix a few compiler errors. on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Remove stale comment. on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Rebased to viable/strict. on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34521686).

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ysiraichi committed:
Invoke constructor explicitly when translating. on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ezyang committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34521686).

ezyang commented:
looks like some more build failures to fix

ezyang commented on aten/src/ATen/native/TensorShape.h:
```diff
@@ -3,6 +3,9 @@
 
 namespace at {
 namespace native {
+static bool cat_should_skip_tensor(const Tensor& t) {
```
Don't mark this static; inline is better.

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -28,6 +29,136 @@
+        canCast(out_dtype, result.scalar_type()),
+        "torch.cat(): input types can't be cast to the desired output type ",
+        result.scalar_type());
+    out_dtype = result.scalar_type();
```
Nothing you need to do for this PR, but this makes me wonder if we should also apply the https://github.com/pytorch/pytorch/issues/69813 treatment to dtype promotion/conversion too... I guess it would have to work differently though!

ysiraichi replied:
Yeah. I think that it would be nice to have that for dtypes, too. I guess, in the end, it would be something similar to what `TensorIterator` kind of already does.

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -28,6 +29,136 @@
+      valid = i;
+      break;
+    }
+  }
```
Man the fact that `[0]` size tensors don't participate in size calculation is so janky. We should deprecate this and then make it just raise an error haha, then we could simplify this code quite a bit. (Nothing you need to do for this PR; if you agree maybe file an issue.)

ezyang reviewed (approved):
I'm satisfied

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ysiraichi committed:
Rename 'ITensorList' to 'ITensorListRef'. on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ysiraichi commented:
I have mistakenly merged this PR with #69606. I will revert it as it was, tomorrow. Sorry.

ysiraichi committed:
Unmerge `cat` and `Tensor?[]` support (mistakenly removed). on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ysiraichi commented:
Everything should be back to normal!

ezyang committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ysiraichi commented:
@ngimel @ezyang @bdhirsh 
I ran some benchmarks on different branches in this stack of PRs:
- `bc512253d5`: the baseline
- [#69607](https://github.com/pytorch/pytorch/pull/69607): similar to this PR, but generalizes `ITensorListRef` to `IListRef`
- [#73351](https://github.com/pytorch/pytorch/pull/73351): uses `IListRef` in the dispatcher, eliminating wasteful copies and ref-count bumps


| Size        | Branch                       | Time CPU (us)                                                    | Time CUDA (us)                            | Instructions                                                                     |
|-------------|------------------------------|----------------------------------------------------------------------|----------------------------------------------------|----------------------------------------------------------------------------------|
| [1,2]       | `bc512253d5`<br>#69607<br>#73351 | 97.01 (2.28)<br>105.14 (3.65)<br>93.15 (4.09)                        | 356.75 (70.23)<br>330.70 (6.69)<br>326.67 (3.44)   | 68323495.50 (12375.54)<br>74569195.40 (12037.75)<br>75222670.40 (20938.94)       |
| [1,2,2]     | `bc512253d5`<br>#69607<br>#73351 | 102.01 (1.11)<br>106.51 (4.32)<br>94.33 (0.86)                       | 347.10 (18.40)<br>363.83 (66.13)<br>361.96 (73.93) | 71919045.40 (2468.85)<br>78584658.80 (17226.44)<br>79226327.90 (20665.40)        |
| [3,256,256] | `bc512253d5`<br>#69607<br>#73351 | 224157.47 (27083.82)<br>239175.47 (36902.33)<br>235462.35 (34238.07) | 2970.30 (6.33)<br>2967.31 (8.96)<br>2968.43 (1.98) | 12363011366.90 (2486.03)<br>12363671642.80 (3776.23)<br>12356348936.20 (2990.37) |

<details>
<summary>Benchmark</summary>
<p>

```python
ntensors = 1000

sizes = [
    (1, 2),
    (1, 2, 2),
    (3, 256, 256),
]

results = []
results_callgrind = []

for size in sizes:
    size_str = f"""[{",".join(str(s) for s in size)}]"""
    timer = benchmark.Timer(
        stmt="torch.cat(xs)",
        setup=f"import torch; xs = [torch.rand(*{size}) for _ in range({ntensors})];",
        label="Cat",
        sub_label=size_str,
        description="time (cpu)"
    )

    timer_cuda = benchmark.Timer(
        stmt="torch.cat(xs); torch.cuda.synchronize()",
        setup=f"import torch; xs = [torch.rand(*{size}, device='cuda') for _ in range({ntensors})]; torch.cuda.synchronize()",
        label="Cat",
        sub_label=size_str,
        description="time (cuda)"
    )

    results.append(timer.blocked_autorange(min_run_time=1))
    results.append(timer_cuda.blocked_autorange(min_run_time=1))
    results_callgrind.append(timer.collect_callgrind())

compare = benchmark.Compare(results)
compare.print()

for r in results_callgrind:
    print(r)
```

</p>
</details>

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ezyang commented:
Since you've got the benchmark, I wonder if you can run a little experiment, which is to try NOT materializing in the body of the kernel, and being willing to iterate multiple times. It may be that the dynamic allocation is swamping the predictable branches, and so we'd rather pump up the instruction count and avoid the dynamic alloc. Would be nice to know one way or another.

ysiraichi committed:
Update on "Port `cat` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D34521686](https://our.internmc.facebook.com/intern/diff/D34521686)

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge this

ezyang commented:
this is about to get yanked from the diff train because it breaks internal users. I just need to update call sites to not use native::

mehtanirav commented:
Fixed the internal references and few other linter errors in [internal diff stack](D35687049)

ysiraichi commented:
> What does the parenthesized number mean?

It's the standard deviation. I left it there just to give a sense of how much the runs varied.

> it seems like no materialize is better despite higher instruction count (what I suspected) and we should use that.

Agreed.
